rpc/src/client/subscription.rs

```diff
@@ -0,0 +1,206 @@
+//! Subscription- and subscription management-related functionality.
+
+use futures::{
+    task::{Context, Poll},
+    Stream,
+};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+use tokio::{stream::StreamExt, sync::mpsc, task::JoinHandle};
+
+use crate::{client::transport::EventConnection, events::Event, Error};
+
+/// The subscription manager is just an interface/handle to the subscription
+/// router, which runs asynchronously in a separate process.
+#[derive(Debug)]
+pub struct SubscriptionManager {
+    router: JoinHandle<Result<(), Error>>,
+    cmd_tx: mpsc::Sender<RouterCmd>,
+    next_subs_id: AtomicUsize,
+}
+
+/// An interface that can be used to asynchronously receive events for a
+/// particular subscription.
+#[derive(Debug)]
+pub struct Subscription {
+    id: SubscriptionId,
+    query: String,
+    event_rx: mpsc::Receiver<Event>,
+}
+
+// The subscription router does the heavy lifting of managing subscriptions and
+// routing incoming events to their relevant subscribers.
+struct SubscriptionRouter {
+    conn: EventConnection,
+    cmd_rx: mpsc::Receiver<RouterCmd>,
+    // Maps queries -> (maps of subscription IDs -> channels)
+    subscriptions: HashMap<String, HashMap<SubscriptionId, mpsc::Sender<Event>>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct SubscriptionId(usize);
+
+#[derive(Debug)]
+enum RouterCmd {
+    Subscribe {
+        id: SubscriptionId,
+        query: String,
+        event_tx: mpsc::Sender<Event>,
+    },
+    Unsubscribe(Subscription),
+    Terminate,
+}
+
+impl SubscriptionManager {
+    pub(crate) fn new(conn: EventConnection, cmd_buf_size: usize) -> Self {
```

Comment:
```suggestion
    pub fn new(conn: EventConnection, cmd_buf_size: usize) -> Self {
```

Unless explicirely reexported this method would be visible outside of the client mod.